File: Tutorials/Codes/Robot_Control_Course/Chapter5/Lesson1/Chapter5_Lesson1.py

import numpy as np

# Robot parameters
l1, l2 = 1.0, 0.7
q1_min, q1_max = -np.pi/2, np.pi/2
q2_min, q2_max = -np.pi, np.pi

# Obstacle parameters
x_obs, y_obs = 0.7, 0.3
r_obs = 0.2

def forward_kinematics(q):
    q1, q2 = q
    x = l1 * np.cos(q1) + l2 * np.cos(q1 + q2)
    y = l1 * np.sin(q1) + l2 * np.sin(q1 + q2)
    return np.array([x, y])

def in_joint_limits(q):
    q1, q2 = q
    return (q1_min <= q1 <= q1_max) and (q2_min <= q2 <= q2_max)

def obstacle_constraint(q):
    x, y = forward_kinematics(q)
    dx = x - x_obs
    dy = y - y_obs
    # h(x) = ||x - x_obs||^2 - r_obs^2 >= 0 in free space
    return dx*dx + dy*dy - r_obs**2

def ground_constraint(q):
    # Non-penetration: y(q) >= 0
    _, y = forward_kinematics(q)
    return y  # should be >= 0

def admissible(q):
    """Check if q satisfies joint, workspace, and ground constraints."""
    if not in_joint_limits(q):
        return False
    if obstacle_constraint(q) < 0.0:
        return False
    if ground_constraint(q) < 0.0:
        return False
    return True

if __name__ == "__main__":
    q_test = np.array([0.0, 0.0])
    print("q_test admissible?", admissible(q_test))

File: Tutorials/Codes/Robot_Control_Course/Chapter5/Lesson1/Chapter5_Lesson1.cpp

#include <iostream>
#include <cmath>
#include <Eigen/Dense>

using Eigen::Vector2d;

struct PlanarArm {
    double l1{1.0};
    double l2{0.7};
    double q1_min{-M_PI/2.0}, q1_max{M_PI/2.0};
    double q2_min{-M_PI},      q2_max{M_PI};
    double x_obs{0.7}, y_obs{0.3}, r_obs{0.2};

    Vector2d forwardKinematics(const Vector2d& q) const {
        double q1 = q(0), q2 = q(1);
        double x = l1 * std::cos(q1) + l2 * std::cos(q1 + q2);
        double y = l1 * std::sin(q1) + l2 * std::sin(q1 + q2);
        return Vector2d(x, y);
    }

    bool inJointLimits(const Vector2d& q) const {
        double q1 = q(0), q2 = q(1);
        return (q1_min <= q1 && q1 <= q1_max) &&
               (q2_min <= q2 && q2 <= q2_max);
    }

    double obstacleConstraint(const Vector2d& q) const {
        Vector2d ee = forwardKinematics(q);
        double dx = ee(0) - x_obs;
        double dy = ee(1) - y_obs;
        return dx*dx + dy*dy - r_obs*r_obs; // should be >= 0
    }

    double groundConstraint(const Vector2d& q) const {
        Vector2d ee = forwardKinematics(q);
        return ee(1); // y >= 0
    }

    bool admissible(const Vector2d& q) const {
        if (!inJointLimits(q)) return false;
        if (obstacleConstraint(q) < 0.0) return false;
        if (groundConstraint(q) < 0.0) return false;
        return true;
    }
};

int main() {
    PlanarArm arm;
    Vector2d q(0.0, 0.0);
    std::cout << "Admissible? " << std::boolalpha
              << arm.admissible(q) << std::endl;
    return 0;
}
